### ISSUE TITLE
[Selection] The selected text in RichText widget is not in correct order when a WidgetSpan is included

### ISSUE THREAD

**Issue by letungcntt** [flutter/flutter#153900]
### Steps to reproduce

1. Launch app sample
2. Drag selection text on screen
3. Copy and paste to another application

### Expected results

We can selet all text by drag mouse

### Actual results

Can't selet all text by drag mouse. TextSpan before WigdetSpan can't selected.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Map> data = [
    {
      'text': "Hello world",
      'number': 1
    }
  ];
  int iHover = 0;

  void _incrementCounter() {
    setState(() {
      data.add({
        'text': "Hello world",
        'number': data.last['number'] + 1
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: SelectionArea(
        child: ListView.builder(
          itemCount: data.length,
          itemBuilder: (context, index) {
            return InkWell(
              hoverColor: Colors.red,
              onTap: () {},
              child: Container(
                color: iHover == data[index]['number'] ? Colors.red : Colors.transparent,
                child: RenderChild(
                  text: data[index]['text'],
                  number: data[index]['number'],
                ),
              ),
            );
          },
        )
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

class RenderChild extends StatefulWidget {
  const RenderChild({
    super.key,
    required this.text,
    required this.number
  });

  final int number;

  final String text;

  @override
  RenderChildState createState() => RenderChildState();
}

class RenderChildState extends State<RenderChild> {
  @override
  Widget build(BuildContext context) {
    return Text.rich(
      TextSpan(
        children: [
          const TextSpan(
            text: 'Hi'
          ),
          WidgetSpan(
            alignment: PlaceholderAlignment.middle,
            child: Container(
              padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4).add(const EdgeInsets.only(bottom: 1)),
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(30),
                color: const Color(0xff72BAFC).withOpacity(0.25)
              ),
              child: Text(
                widget.number.toString(),
                style: const TextStyle(
                  color: Color(0xff1677F7),
                  fontWeight: FontWeight.w500,
                ),
              )
            )
          ),
          TextSpan(
            // text: widget.text,
            text: '${widget.text} Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum'
          )
        ]
      )
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/user-attachments/assets/0fe7ec45-cd88-44b8-a7cc-69f681cc4400

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.22.3, on macOS 14.6.1 23G93
    darwin-arm64, locale en-VN)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from:
      https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android
      SDK components.
      (or visit
      https://flutter.dev/docs/get-started/install/macos#android-set
      up for detailed instructions).
      If the Android SDK has been installed to a custom location,
      please use
      `flutter config --android-sdk` to update to that location.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
[✓] Chrome - develop for the web
[!] Android Studio (not installed)
[✓] VS Code (version 1.92.2)
[✓] Connected device (4 available)
[✓] Network resources

! Doctor found issues in 2 categories.
```

</details>

**Comment by letungcntt** [comment]
I have check https://github.com/flutter/flutter/issues/127942. if text after WidgetSpan dont push new line. the problem not existed

**Comment by huycozy** [comment]
Hi @letungcntt, I checked this on Flutter 3.24.1 but didn't see the issue. Can you try upgrading SDK and confirm?

https://github.com/user-attachments/assets/0d663c76-797c-44cb-9f3c-2cd7ebb638eb

**Comment by moffatman** [comment]
Text -> widgetspan -> Text in same line has some broken selection orders

TO future flutter assignee: I have various selection fixes in my own branch https://github.com/moffatman/flutter/commits/master/

It may be this fix
https://github.com/moffatman/flutter/commit/67d9205e8cd73a90dbca99ee71c89ed6d1f9f83a

**Comment by letungcntt** [comment]
> Hi @letungcntt, I checked this on Flutter 3.24.1 but didn't see the issue. Can you try upgrading SDK and confirm?
> 
>  Screen.Recording.2024-08-22.at.18.02.23.mov

@huycozy Thanks for solution. I confirm the problem has fixed with 3.24.1

**Comment by huycozy** [comment]
Thanks for confirming this. Closing the issue as fixed on the latest stable.
